### PR TITLE
Made json encoding/decoding configurable + Aeson compatible encoding/decoding

### PR DESCRIPTION
**Pull request by eskimor**
in order to provide an aeson compatible encoding/decoding option. Also resolving issue #5.

### CODE REVIEW

**Comment by garyb** [comment]
@hdgarrood could you take a look at this perhaps, if you have time? You're definitely more familiar with Aeson than me. :smile:

File: src/Data/Argonaut/Options.purs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+aesonSumEncoding :: SumEncoding
+aesonSumEncoding = TaggedObject {
+  tagFieldName           : "tag"
+, contentsFieldName      : "contents"
+}
+
+
+
+allConstructorsNullary :: Array DataConstructor -> Boolean
+allConstructorsNullary constrSigns = foldr (&&) true <<< map (\c -> null c.sigValues) $ constrSigns
```
Comment (hdgarrood) [inline src/Data/Argonaut/Options.purs]: Is `foldr (&&) true` the same as `Data.Foldable.all`?

**Comment by eskimor** [comment]
I tried to make it backwards compatible. So gEncodeJson became a simple wrapper around genericEncodeJson which takes an Options parameter resulting in the same format previously used.
Breaking backwards compatibility is only, that I dropped gEncodeJson' for obvious reasons, which was exported, but I believe only for testing.

File: src/Data/Argonaut/Options.purs (hunk 1 above)
Comment (hdgarrood) [inline src/Data/Argonaut/Options.purs]: Ok, yeah, I think this could be written `allConstructorsNullary = Data.Foldable.all (null <<< _.sigValues)`

Comment (eskimor) [inline src/Data/Argonaut/Options.purs]: Yeah - I missed Data.Foldable.all during my Pursuit search. Thanks!

**Comment by eskimor** [comment]
Btw. guys - you are amazing. Sending PRs to purescript projects is so much more fun, than for nixos. Answers in seconds, instead of weeks/months. I don't know how you do that, but it is really motivating.

**Comment by hdgarrood** [comment]
This looks good to me. Regarding Aeson specifically, I'd just say, if it works, go ahead :)

**Comment by hdgarrood** [comment]
I do it by happening to be on github at the same time as you, and not having many other commitments right now :smile: it is nice to know that it's appreciated, though. Incidentally, that (not having many commitments) probably won't remain the case for very much longer for me.

**Comment by garyb** [comment]
@hdgarrood congrats? :wink:

@eskimor I'm in the lucky position to work with PureScript full time now, and especially for libraries like this that we (SlamData) use it means I try to stay on top of things. I don't always manage it though!

**Comment by garyb** [comment]
Code looks fine to me too, :+1: but I'd appreciate it if we could have at least a few tests for the Aeson side of things before merging, perhaps just a basic roundtrip encode/decode test for each sig case?

**Comment by hdgarrood** [comment]
@garyb It's maybe a little early for that, but cheers anyway :)

**Comment by eskimor** [comment]
@garyb : Done :-) I also checked the output - it looks good and aeson compatible. 
Is quickCheck generating good test data? It looks too good to be true.

**Comment by garyb** [comment]
It depends how good the generators are :smile: but yep, it's wonderful for this kind of testing!

Thanks :+1:

**Comment by garyb** [comment]
If you don't mind I'll hold off merging this until we do the big psc 0.8 library update (should be really soon now!), as technically this is a breaking change (dropping the primed exports) so will mean bumping all the dependants too. We can add the `unsafeCrashWith` cases in then too.

**Comment by eskimor** [comment]
Ok - no problem, I will use my local version in the meantime.

**Comment by eskimor** [comment]
Glad have not merged it yet - I added two more tests and indeed discovered that it was not quite right yet. Ouch! Ouch! Ouch! It was too good to be true.

**Comment by eskimor** [comment]
Resolved issue #5.

**Comment by eskimor** [comment]
Hmm - the tests passed for me locally - I assume the testsuite does not use psc 0.8 yet?

**Comment by eskimor** [comment]
Bumped psc version in package.json to 0.8.0 - incorporated the unsafeCrashWith stuff.

**Comment by eskimor** [comment]
Hi there! Can this be merged now? Or is something still missing?

**Comment by garyb** [comment]
We haven't finished updating the core libraries for 0.8 yet. I was intending to wait until then, as we'll need to bump all the dependency versions again when that's finished, which will be classed as another breaking change.

**Comment by eskimor** [comment]
Ah I get it now, you would release a new version of purescript-argonaut-codecs too when merging? So master and the latest version always stay the same? I see, ping me if anymore changes are needed for merging.

**Comment by eskimor** [comment]
Resolved issue #14

**Comment by eskimor** [comment]
Huh - what's wrong with the test suite?

**Comment by eskimor** [comment]
Creating standalone [package](https://github.com/eskimor/purescript-argonaut-generic-codecs).

**Pull request closed without merging**